Dataflow based differ implementation

vish-cs opened the pull request.

n-h-diaz reviewed (commented):
Thanks for adding this!

It would be helpful if you added a short readme on usage and how to understand the output

Also was just wondering: 
- Are there any opportunities for integration/sharing with existing import tool? (Particularly around mcf parsing/utils - if I understand correctly, this was why we went with Java Dataflow) 
- Is there any plan to add time series-level stats? (It looks like currently it's just on the obs level)

vish-cs commented on pipeline/differ/src/main/java/org/datacommons/DifferPipeline.java:
```diff
@@ -0,0 +1,47 @@
+
+    // Read input graph files and convert into PCollections.
+    PCollection<McfGraph> previous_nodes, current_nodes;
+    if (options.getUseTfrFormat().isAccessible()) {
```
It could potentially be. The current implementation does not support that. We can add that if needed.

vish-cs commented on pipeline/differ/src/main/java/org/datacommons/DifferUtils.java:
```diff
@@ -0,0 +1,163 @@
+  public static PCollection<McfGraph> readMcfGraph(ValueProvider<String> files, Pipeline p) {
+    return p.apply(
+            "ReadMcfGraph", TFRecordIO.read().from(files).withCompression(GZIP).withoutValidation())
+        // .apply(ParDo.of(new ByteArrayToString()));
```
Deleted

vish-cs commented on pipeline/differ/src/test/java/org/datacommons/DifferTest.java:
```diff
@@ -0,0 +1,50 @@
+
+  @Test
+  public void testDiffer() {
+    options.setStableUniqueNames(PipelineOptions.CheckEnabled.OFF);
```
The test complains otherwise since we run the same pipeline stage with the same name two times (previous and current Data).

vish-cs commented on pipeline/util/src/main/java/org/datacommons/GraphUtil.java:
```diff
@@ -0,0 +1,226 @@
+    String value = GetPropertyValue(pvs, Property.value.name());
+    String scaling;
+    if (!(scaling = GetPropertyValue(pvs, Property.scalingFactor.name())).isEmpty()) {
+      value = value + "*" + scaling;
```
Ajai and I discussed this. We can combine scaling factor with the value since it's basically a multiple for the value. No?

n-h-diaz replied:
Practically it is, but internally we separate it out into different time series https://github.com/datacommonsorg/mixer/blob/ae18cc9dac359342c0aa0388c768fbceee726dec/proto/stat.proto#L40  so multiplying them could hide this distinction. Is there any reason to combine them? @ajaits 
 Otherwise for consistency it might be helpful to have them separate

ajaits replied:
scaling_factor needs to be divided from the SVObs value to get the raw value. 

Based on other facet discussions, i think it may be worth preserving the facet definition of a timeseries in the diff too, i.e., make scaling_factor part of the key. We may formalize facet definition into the schema for other uses.

vish-cs replied:
I have updated the PR to include scaling_factor in the key

vish-cs commented on pipeline/differ/src/main/java/org/datacommons/DifferUtils.java:
```diff
@@ -0,0 +1,163 @@
+    ADDED,
+    DELETED,
+    MODIFIED,
+    TOTAL,
```
It was mostly used for testing/debugging etc. I have removed it.

vish-cs commented on pipeline/differ/src/main/java/org/datacommons/DifferUtils.java:
```diff
@@ -0,0 +1,163 @@
+    PCollection<KV<String, CoGbkResult>> joinedResult =
+        KeyedPCollectionTuple.of(keyTag, currentNodes)
+            .and(valueTag, previousNodes)
+            .apply(CoGroupByKey.create());
```
Makes sense. I have renamed to current/previous tag

vish-cs commented on pipeline/util/src/main/java/org/datacommons/GraphUtil.java:
```diff
@@ -0,0 +1,226 @@
+      SetPropVal(lhs, ValueType.TEXT, rhs, base_node);
+    }
+    McfGraph.PropertyValues pvs = base_node.build();
+    if (GetPropertyValue(pvs.getPvsMap(), Property.typeOf.name())
```
That's right. I have updated the function description to clarify.

vish-cs commented on pipeline/util/src/main/java/org/datacommons/GraphUtil.java:
```diff
@@ -0,0 +1,226 @@
+      SetPropVal(Property.scalingFactor.name(), ValueType.NUMBER, m, base_node);
+    }
+    if (!(m = s_key.getUnit()).isEmpty()) {
+      SetPropVal(Property.unit.name(), ValueType.TEXT, m, base_node);
```
We are just concatenating all the property values as key and diff for that node as value so we don't need the resolved_ref.

vish-cs commented on pipeline/differ/src/main/java/org/datacommons/DifferOptions.java:
```diff
@@ -0,0 +1,39 @@
+  void setUseTfrFormat(ValueProvider<String> value);
+
+  @Description("Whether to use mcf file format")
+  ValueProvider<String> getUseMcfFormat();
```
I was also debating how to best configure these flags. I have changed it to use a single flag useOptimizedGraphFormat

vish-cs commented:
> Thanks for adding this!
> 
> It would be helpful if you added a short readme on usage and how to understand the output
I have added a README in the code.

> 
> Also was just wondering:
> 
> * Are there any opportunities for integration/sharing with existing import tool? (Particularly around mcf parsing/utils - if I understand correctly, this was why we went with Java Dataflow)
Actually that's the idea. GraphUtils can be a part of the common libs. I did not find any existing functions that I could resuse so I have added new routines. I have kept them under pipeline module for now but it can be moved.  

> * Is there any plan to add time series-level stats? (It looks like currently it's just on the obs level)
Actually we already have time-series stats. The idea is that this pipeline generates the raw diff which gets processed in a python script to generate obs-level and series-level stats

n-h-diaz commented on pipeline/differ/src/main/java/org/datacommons/README.md:
```diff
@@ -0,0 +1,21 @@
+
+This module implements a Dataflow pipeline that generates a summary of changes between two versions (current and previous) of the same dataset. It does that by performing an outer join on the datasets. It supports both MCF and TFRecord file formats as input. The generated output (CSV) includes the following information:
+
+- Stat var: string containing the property values in the observation
```
Could you mention what these fields are/their order in the key?

n-h-diaz commented:
> > Thanks for adding this!
> > It would be helpful if you added a short readme on usage and how to understand the output
> > I have added a README in the code.
> 
> > Also was just wondering:
> > 
> > * Are there any opportunities for integration/sharing with existing import tool? (Particularly around mcf parsing/utils - if I understand correctly, this was why we went with Java Dataflow)
> >   Actually that's the idea. GraphUtils can be a part of the common libs. I did not find any existing functions that I could resuse so I have added new routines. I have kept them under pipeline module for now but it can be moved.
> 
> > * Is there any plan to add time series-level stats? (It looks like currently it's just on the obs level)
> >   Actually we already have time-series stats. The idea is that this pipeline generates the raw diff which gets processed in a python script to generate obs-level and series-level stats

Does the python script already exist? Just curious why it's split into two stages vs doing the summary stats here too (sorry if I'm just missing a lot of context on this - I'm not too familiar with all the various import tools)

n-h-diaz reviewed (commented):
Thanks for the updates! I think this generally looks good, just wanted to clarify how we're handling scaling_factor 

(cc @ajaits , in can you have any thoughts on this!)

vish-cs commented:
> > > Thanks for adding this!
> > > It would be helpful if you added a short readme on usage and how to understand the output
> > > I have added a README in the code.
> > 
> > 
> > > Also was just wondering:
> > > 
> > > * Are there any opportunities for integration/sharing with existing import tool? (Particularly around mcf parsing/utils - if I understand correctly, this was why we went with Java Dataflow)
> > >   Actually that's the idea. GraphUtils can be a part of the common libs. I did not find any existing functions that I could resuse so I have added new routines. I have kept them under pipeline module for now but it can be moved.
> > 
> > 
> > > * Is there any plan to add time series-level stats? (It looks like currently it's just on the obs level)
> > >   Actually we already have time-series stats. The idea is that this pipeline generates the raw diff which gets processed in a python script to generate obs-level and series-level stats
> 
> Does the python script already exist? Just curious why it's split into two stages vs doing the summary stats here too (sorry if I'm just missing a lot of context on this - I'm not too familiar with all the various import tools)

Yes the script exists here:
https://github.com/datacommonsorg/data/pull/1243

The reason for splitting is to have the core data processing logic in dataflow while the dependent aggregation operations since that logic complicated the datalow pipeline (point analysis, series analysis etc.) Also, it keeps the generic differ pipeline generic which can feed into other components like incremental ingestion.
